plugins/android_boot/fu-android-boot-device.c

```diff
@@ -0,0 +1,441 @@
+		g_debug("Partition UUID: %s", self->uuid);
+	}
+
+	/* set max firmware size, required to avoid writing firmware bigger than partition */
+	if (g_udev_device_has_property(udev_device, "ID_PART_ENTRY_SIZE")) {
+		sectors = g_udev_device_get_property_as_uint64(udev_device, "ID_PART_ENTRY_SIZE");
+		size = sectors * SECTOR_SIZE;
+		g_debug("Partition size: %" G_GUINT64_FORMAT " bytes", size);
+		self->max_size = size;
+	} else {
+		g_set_error_literal(error,
+				    FWUPD_ERROR,
+				    FWUPD_ERROR_NOT_SUPPORTED,
+				    "device does not expose its size");
+		return FALSE;
+	}
+
+	/* support all partitions with a UUID */
+	if (self->uuid == NULL) {
+		g_set_error_literal(error,
+				    FWUPD_ERROR,
+				    FWUPD_ERROR_NOT_SUPPORTED,
+				    "device not supported");
+		return FALSE;
+	}
+
+	/* reject partitions from an inactive slot in an A/B scheme */
+	if (self->boot_slot != NULL && !matches_boot_slot) {
+		g_set_error_literal(error,
+				    FWUPD_ERROR,
+				    FWUPD_ERROR_NOT_SUPPORTED,
+				    "device is on a different bootslot");
+		return FALSE;
+	}
+
+	/* set serial */
+	serial = fu_android_boot_extract_cmdline_arg_value(SERIAL_ARG);
+	fu_device_set_serial(FU_DEVICE(self), serial);
+
+	/*
+	 * Some devices don't have unique TYPE UUIDs, add the partition label to make them truly
+	 * unique Devices have a fixed partition scheme anyway because they originally have Android
+	 * which has such requirements.
+	 */
+	fu_device_add_instance_strsafe(device, "UUID", self->uuid);
+	fu_device_add_instance_strsafe(device, "LABEL", self->label);
+	fu_device_add_instance_strsafe(device, "SLOT", self->boot_slot);
+
+	/* GUID based on UUID */
+	fu_device_build_instance_id(device, error, "DRIVE", "UUID", NULL);
+
+	/* GUID based on label, and UUID */
+	fu_device_build_instance_id(device, error, "DRIVE", "UUID", "LABEL", NULL);
+
+	/* GUID based on label, boot slot, and UUID */
+	fu_device_build_instance_id(device, error, "DRIVE", "UUID", "LABEL", "SLOT", NULL);
+
+	/* Success */
+	return TRUE;
+}
+
+static gboolean
+fu_android_boot_device_setup(FuDevice *device, GError **error)
+{
+	FuAndroidBootDevice *self = FU_ANDROID_BOOT_DEVICE(device);
+
+	/*
+	 * Fallback for ABL without version reporting, fwupd will always provide an upgrade in this
+	 * case. Once upgraded, the version reporting will be available and the update notification
+	 * will disappear. If version reporting is available, the reported version is set.
+	 */
+	fu_device_set_version(FU_DEVICE(self),
+			      self->version == NULL ? UNKNOWN_VERSION : self->version);
+
+	fu_device_set_firmware_size_max(device, self->max_size);
+
+	/* Success */
+	return TRUE;
+}
+
+static gboolean
+fu_android_boot_device_open(FuDevice *device, GError **error)
+{
+	g_autoptr(GError) error_local = NULL;
+
+	/* FuUdevDevice->open */
+	if (!FU_DEVICE_CLASS(fu_android_boot_device_parent_class)->open(device, &error_local)) {
+		if (g_error_matches(error_local, G_IO_ERROR, G_IO_ERROR_PERMISSION_DENIED)) {
```

Comment:
how exactly does this happen?

Reply:
Most of the writing, reading, erasing comes from the MTD plugin.
I re-used this part as well, in theory you can get a permission denied for devices which are not readable or something something with AppArmor/Flatpak/SELinux stuff. This is mainly a precaution I think?